### ISSUE TITLE
DRPolicy reconciler to install dr-cluster components on managed clusters

### ISSUE THREAD

**Issue by hatfieldbrian** [RamenDR/ramen#335]
# Existing implementation

[Ramen installation](https://github.com/RamenDR/ramen/blob/main/docs/install.md) steps include:
- deploy ramen hub operator and custom resource definitions
- deploy ramen cluster operator and custom resource definitions on each managed cluster

[Ramen configuration](https://github.com/RamenDR/ramen/blob/main/docs/configure.md) steps include:
- create ramen config map and s3 profile(s) and secret(s) on hub (for drpolicy validation)
- create ramen config map and s3 profile(s) and secret(s) on each managed cluster

# Objectives

The primary objective of this issue is to make deployment easier such that an administrator need only interact with a hub to install and configure ramen.  A ramen hub operator deploys and undeploys resources on managed clusters.

- Hub deploys ramen dr-cluster operator to managed clusters
- Hub undeploys ramen-dr-cluster operator from managed clusters when no longer needed
- Hub replicates its ramen config map and s3 secrets to managed clusters including updates and deletions
- Hub selectively replicates s3 profiles and their secrets to managed clusters that require them
- Hub s3 secrets are encrypted for transfer to managed clusters and decrypted on receipt
- Hub reports status of managed cluster operators?

Issue https://github.com/RamenDR/ramen/issues/102 specifically tracks deployment of s3 profiles and secrets to managed clusters.

## drpolicy

Managed cluster names and their s3 profile references are specified in a hub's drpolicies.  Therefore it has been proposed that the drpolicy reconciler be responsible for managed cluster resource deployment.

### Complexities and inefficiencies

One issue with this approach is that a given managed cluster may be specified in multiple drpolicies.  Creating a managed cluster resource that already exists typically does not cause a problem.  However, a managed cluster resource should only be removed once no drpolicies specify it.  Serialization can be used to resolve races among concurrent drpolicy reconciliations.  A set data structure can be used to determine which clusters are contained in drpolicies.

These complexities (serialization, logic to determine if a cluster is referenced by any drpolicies) and inefficiencies (duplicating work for clusters found in multiple drpolicies), could be avoided if dr managed clusters were specified in an object which has only one instance per cluster, perhaps, ramen config or some new drClusterSet custom resource and an associated reconciler.  A drpolicy could then reference a subset of clusters specified in such an object, though I'm not aware of any requirement for that at this point.

At this point the intent is to extend the drpolicy and manage the complexities and inefficiencies associated with it.

## Implementation

open-cluster-management has at least 3 ways to deploy resources on a managed cluster:
- [manifest work](https://open-cluster-management.io/concepts/manifestwork/)
- [policy](https://open-cluster-management.io/getting-started/integration/policy-framework/)
- [addon](https://open-cluster-management.io/concepts/addon/)

### manifest work
This is the current method used to deploy cluster roles and cluster role bindings and could be extended to include other resources.  one downside to this approach is that once a manifest work is deleted there is no longer an object to report status of whether the resources were successfully removed from the managed cluster or not.  Though if I recall correctly, Joshua Packer said that release 2.5 is supposed to fix  that allowing the "payload" of the manifest work to be deleted leaving a place to report status and perhaps using "reverse proxy".  A solution to this issue that is available now is to poll for managed cluster resource absence using a managed cluster view.

### policy
This approach is similar to an application deployment where manifests are stored in a git or hlem repo or object store and a subscription and placement rule with cluster selector are employed.  It requires managed clusters labels be maintained which is considered a drawback.  Additionally, it is preferred that the internet need only be accessed once to install the hub operator from quay.io and not not again access say github.com to deploy dr-cluster resources.

### addon
This approach is, according to Joshua Packer, a more integrated solution and the method being pursued.  There are several implementations of it including:
- ocm's [helloworld](https://github.com/open-cluster-management-io/addon-framework/tree/main/examples/helloworld) example
- [submariner](https://github.com/open-cluster-management/submariner-addon)
- odf multicluster orchestrator's [token exchanger](https://github.com/red-hat-storage/odf-multicluster-orchestrator/tree/main/addons/token-exchange)

Each of these examples uses:
- golang's [embed](https://pkg.go.dev/embed) package to import the contents of yaml files into an operator at compile time where each yaml file contains one object of kinds including deployments, roles, role bindings, and service accounts
- {{ .variables }} in yaml files to specify things like:
  - name of managed cluster namespace to deploy resources in
  - name of container image

Ramen, in contrast uses kustomize to:
- specify namespace and container image names
- produce one yaml file per operator containing several objects

Additionally ramen yaml manifest files are located in its `config` directory and its controllers' go sources are contained in its `controllers` directory.  However, the [embed](https://pkg.go.dev/embed#hdr-Directives) package supports files in the source's file directory or a subdirectory:
> A //go:embed directive above a variable declaration specifies which files to embed, using one or more path.Match patterns.

> The patterns are interpreted relative to the package directory containing the source file.

> Patterns may not contain ‘.’ or ‘..’

So, the addon code must be located in say ramen repo's root directory or its `config` directory.  Or yaml manifest files must be located  elsewhere, say the `controllers` directory or a new `addons` directory.

**Comment by hatfieldbrian** [comment]
We decided to:
1. Use the existing, known, manifest work method to deploy resources to the managed clusters.  According to the latest [documentation](https://open-cluster-management.io/concepts/manifestwork/), deleting a manifest work should result in the corresponding managed cluster objects being deleted in the `Foreground` and presumably only once their deletion is finalized, finalizing the manifest work deletion.
1. Deploy the olm resources `CatalogSource`, `OperatorGroup`, and `Subscription` to the clusters in a drpolicy.  This results in deployment of the dr-cluster olm bundle including the dr-cluster operator, volume resource group custom resource definition, and a configmap containing/referencing the ramen config.

This issue now tracks the deployment of the dr-cluster olm bundle and its corresponding end-to-end test.  A unit test was not written due to dependencies on the olm operators.

Issue #102 continues to track drpolicy's configuration of its dr-clusters.
